store/tikv/region_request.go

```diff
@@ -633,14 +647,22 @@ func regionErrorToLabel(e *errorpb.Error) string {
 	return "unknown"
 }
 
-func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, seed *uint32, regionErr *errorpb.Error) (shouldRetry bool, err error) {
+func (s *RegionRequestSender) onRegionError(bo *Backoffer, ctx *RPCContext, seed *uint32, regionErr *errorpb.Error) (shouldRetry bool, opts []StoreSelectorOption, err error) {
 	if span := opentracing.SpanFromContext(bo.GetCtx()); span != nil && span.Tracer() != nil {
 		span1 := span.Tracer().StartSpan("tikv.onRegionError", opentracing.ChildOf(span.Context()))
 		defer span1.Finish()
 		// TODO(MyonKeminta): Make sure trace works without cloning the backoffer.
 		// bo = bo.Clone()
 		bo.SetCtx(opentracing.ContextWithSpan(bo.GetCtx(), span1))
 	}
+	opts = ctx.storeSelectorOptions
+	// Stale Read request will retry the leader or next peer on error,
+	// so we will exclude the PeerID of the requested peer every time.
+	// If the new PeerID keeps being the same with the last one, we
+	// should not continue excluding it to make the opts become bigger.
+	if ctx.isStaleRead && ctx.lastPeerID != ctx.Peer.GetId() {
+		opts = append(opts, WithExcludedPeerIDs([]uint64{ctx.Peer.GetId()}))
+	}
```

Comment:
The fallback strategy is kind of different from what I think, I think if the `tx_scope` is `local`, we shall find other store which located in same dc. If the `tx_scope` is `global`, we shall directly send request to leader.

Reply:
For now, it works like this:

- If the `tx_scope` is local, it will find other stores located in the same DC according to matched labels and try on different peers one by one on the error. If all peers are tried, it will only retry on the leader until fails.
- If the `tx_scope` is global, it will try on different peers one by one on the error. If all peers are tried, it will only retry on the leader until fails.

It seem that we only need to change the behaviour of the `global`? Once it fails, it will directly request to the leader rather than try all other peers.

Reply:
I think it's better to send to leader directly when `tx_scope` is `global`